portal/src/main/scala/cromwell/pipeline/service/UserService.scala

```diff
@@ -1,15 +1,47 @@
 package cromwell.pipeline.service
 
 import cromwell.pipeline.datastorage.dao.repository.UserRepository
+import cromwell.pipeline.datastorage.dto.user.{ PasswordUpdateRequest, UserUpdateRequest }
 import cromwell.pipeline.datastorage.dto.{ UserId, UserNoCredentials }
+import cromwell.pipeline.utils.StringUtils
 
 import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Random }
 
 class UserService(userRepository: UserRepository)(implicit executionContext: ExecutionContext) {
+  import StringUtils._
 
   def deactivateUserById(userId: UserId): Future[Option[UserNoCredentials]] =
     for {
       _ <- userRepository.deactivateUserById(userId)
       user <- userRepository.getUserById(userId)
     } yield user.map(UserNoCredentials.fromUser)
+
+  def updateUser(userId: String, request: UserUpdateRequest): Future[Int] =
+    userRepository
+      .getUserById(UserId(userId))
+      .flatMap(
+        user =>
+          userRepository.updateUser(
+            user.get.copy(email = request.email, firstName = request.firstName, lastName = request.lastName)
+          )
+      )
+
+  def updatePassword(
+    userId: String,
+    request: PasswordUpdateRequest,
+    salt: String = Random.nextLong().toHexString
+  ): Future[Int] =
+    userRepository
+      .getUserById(UserId(userId))
+      .filter(user => user.get.passwordHash == calculatePasswordHash(request.currentPassword, user.get.passwordSalt))
+      .filter(_ => (request.newPassword == request.repeatPassword))
```

Comment:
What is the point of filtering some value if we do not even care about the value itself?
It is better to check whether the new and repeated passwords are equal before an attempt to find a user.

Reply:
Done